api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+package com.nicknathanjustin.streamercontracts.votes;
+
+import com.nicknathanjustin.streamercontracts.contracts.ContractModel;
+import com.nicknathanjustin.streamercontracts.users.UserModel;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import java.sql.Timestamp;
+import java.util.UUID;
+
+@RequiredArgsConstructor
+@Slf4j
+public class VoteServiceImpl implements VoteService{
+
+    @NonNull final VoteModelRepository voteModelRepository;
+
+    @Override
+    public void recordVote(@NonNull final UserModel voter,
+                           @NonNull final ContractModel contractModel,
+                           @NonNull final boolean flaggedCompleted) {
+        if (userCanVoteOnContract(voter, contractModel)) {
+            final VoteModel voteModel = VoteModel.builder()
+                    .voter(voter)
+                    .contract(contractModel)
+                    .viewerFlaggedComplete(flaggedCompleted)
+                    .votedAt(new Timestamp(System.currentTimeMillis()))
+                    .build();
+            voteModelRepository.save(voteModel);
+        }
+    }
+
+    @Override
+    public boolean isVotingComplete(@NonNull final ContractModel contractModel) {
+        validateContract(contractModel);
+
+        final UUID contractId = contractModel.getId();
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getProposer().getId());
+        final VoteModel streamerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getStreamer().getId());
+        final Timestamp now = new Timestamp(System.currentTimeMillis());
+        return (proposerVote != null && streamerVote != null) ||
```

Comment:
Let's add a comment here that carefully explains the logic. This is essentially the core of the voting logic.

Reply:
Its simple now, but that's only because we're handling the non-community contract case. Community contracts will be more complex.

I'm not saying comment explaining each line of code, obviously the code explains that and adding comments there is redundant. What's needed is a comment here explaining the high level logic of what's happening here. Something that explains WHY the code is the way it is, not just WHAT its doing.

I honestly can't tell you how many times I've encountered very readable code but I have no idea why it was coded that way.